test/suite/test_txn27.py

```diff
@@ -0,0 +1,90 @@
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import wiredtiger, wttest, time
+from wtdataset import SimpleDataSet
+
+# test_txn27.py
+#   Test that the API returning a rollback error sets the reason for the rollback.
+class test_txn27(wttest.WiredTigerTestCase):
+    conn_config = 'cache_size=1MB'
+
+    def test_rollback_reason(self):
+        uri = "table:txn27"
+        # Create a very basic table.
+        ds = SimpleDataSet(self, uri, 10, key_format='S', value_format='S')
+        ds.populate()
+
+        # Update key 5 in the first session.
+        session1 = self.session
+        cursor1 = session1.open_cursor(uri)
+        session1.begin_transaction()
+        cursor1[ds.key(5)] = "aaa"
+
+        # Update the same key in the second session, expect a conflict error to be produced.
+        session2 = self.conn.open_session()
+        cursor2 = session2.open_cursor(uri)
+        session2.begin_transaction()
+        cursor2.set_key(ds.key(5))
+        cursor2.set_value("bbb")
+        msg1 = '/conflict between concurrent operations/'
+        self.assertRaisesException(wiredtiger.WiredTigerError, lambda: cursor2.update(), msg1)
+        self.assertEquals('/' + session2.get_rollback_reason() + '/', msg1)
+
+        # Rollback the transactions, check that session2's rollback error was cleared.
+        session2.rollback_transaction()
+        self.assertEquals(session2.get_rollback_reason(), None)
+        session1.rollback_transaction()
+
+        # Start a new transaction and insert a value far too large for cache.
+        session1.begin_transaction()
+        cursor1.set_key(ds.key(1))
+        cursor1.set_value("a"*1024*5000)
+        self.assertEqual(0, cursor1.update())
+
+        # Let WiredTiger's accounting catch up.
+        time.sleep(2)
+
+        # Attempt to insert another value with the same transaction. This will result in the
+        # application thread being pulled into eviction and getting rolled back.
+        cursor1.set_key(ds.key(2))
+        cursor1.set_value("b"*1024)
+
+        # This is the message that we expect to be raised when a thread is rolled back due to
+        # cache pressure. MongoDB expects the same so if this test breaks due to this message
+        # changing then care must be taken to ensure that MongoDB is aware.
+        msg2 = 'oldest pinned transaction ID rolled back for eviction'
+        # Expect stdout to give us the true reason for the rollback.
+        with self.expectedStdoutPattern(msg2):
+            # This reason is the default reason for WT_ROLLBACK errors so we need to catch it.
+            self.assertRaisesException(wiredtiger.WiredTigerError, lambda: cursor1.update(), msg1)
+        # Expect the rollback reason to give us the true reason for the rollback.
+        self.assertEquals(session1.get_rollback_reason(), msg2)
+
+if __name__ == '__main__':
+    wttest.run()
```

Comment:
I agree this should fire, but I worry a future change to the library will break this test and it will be difficult to debug.

Maybe it would be better/safer to create `#define WT_ROLLBACK_REASON_XXX` in the library for each "reason", with the comment you have here with the list of reasons, so it's obvious to anyone changing the code the message shouldn't change, and then we wouldn't need a specific test for each message?

Reply:
I think that solution also has issues as msg1 `'conflict between concurrent operations'` is defined in api_err.py. If we were to #define it then we'd have it in two places (again).

Reply:
Happy to go with your call.